### ISSUE TITLE
New Document > Default Language: add "☐ Apply to unidentified files" option

### ISSUE THREAD

**Issue by pryrt** [notepad-plus-plus/notepad-plus-plus#12256]
### Description of the Request

In the **Settings > Preferences > New Document** interface, add "☐ Apply to unidentified files" option to the Default Language

![](https://user-images.githubusercontent.com/17455758/192624384-9b022bc4-fba9-468a-91d5-70923fdc3f11.png)

see [this comment](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/12236#issuecomment-1259988624) and following for context

**Comment by alankilborn** [comment]
Is that a suggestion for real hover text?  Translators will have some fun with that one! :-)

**Comment by pryrt** [comment]
You're right.  That is way too complicated.  Maybe the hover text should be more like:
> Applies Default Language to files whose language cannot be determined when opened

Or, to be more like the other checkbox on the same settings page, just change the option string to `☐ Apply to unidentified files when opened` , and don't bother with the hover text.

I have never claimed to be good at UI design, and am known for being wordier than I need to be.

The main point of this request is to give users control over whether or not the "Default language" setting will be applied to files that are opened-but-unidentified, or whether "Default language" should only apply to new files.

**Comment by rdipardo** [comment]
> add "☐ Apply to unidentified files" option to the Default Language

Aren't we forgetting something?

Observe what happens to a [Code::Blocks project file][0] when a default language is chosen.

Oddball extension (`*.cbp`) => unknown file type

Result => XML hightlighting!

Lucky guess? Nope. The default language is simply ignored __*anytime*__ the file has __*any kind*__ of extension and __*the contents match a detectable file type*__!

If we're serious about rationalizing all this arbitrary behaviour, let's at least try to cover every case we know about: 


       [ ] Apply to unidentified files -- with an unknown file extension,
           even when the file's content matches a detectable language 
           (i.e. always ignore file content)

       Current default - NO (apparently with no reported complaints)

       [ ] Apply to unidentified files -- with no extension,
           even when the file's content matches a detectable language 
           (i.e. always ignore file content)

       Previous default  - YES (with at least one known complaint in #11504)
       Future default    - probably YES, for historical reasons 

Logically, a single option would cover them all:

       Default Language
       ----------------

       [ ] Always ignore file content


[0]: https://github.com/notepad-plus-plus/notepad-plus-plus/pull/12236#:~:text=Set%20a%20default%20language%20(e.g.%20C%2B%2B)%3B%20open%20the%20same%20Code%3A%3ABlocks%20project%20file

**Comment by ArkadiuszMichalski** [comment]
>Lucky guess? Nope. The default language is simply ignored anytime the file has any kind of extension and the contents match a detectable file type!

This is due to the subsequent stages of detection (extension, content). But at the very end we can add one more condition that if `[ ] default lang` from Prefs is enabled then just use it (if `_lang` is still `L_TEXT`). After this condition or as `else if`:
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/0ce3bc08df1a7de8eff1ec1d70ee3b713224407c/PowerEditor/src/ScintillaComponent/Buffer.cpp#L745

There is no need to modify the already made detection (and risk another regression). Note that this could be turned on / off in Prefs, so if it disturbs someone it can always change it. 

In practice this only modify last step 7. https://github.com/notepad-plus-plus/notepad-plus-plus/pull/12236#issuecomment-1257168802. Keeps your corrections from the last PR. So we have what was before, with one extra check of something that worked halfway anyway (because this default langs was only applied to unknown files without extension, and now it would also apply to a file with the extension).
I remember that there were requests for this default lang from Prefs to be taken into account also for opened files (as last hope, instead of default L_TEXT), not only for new documents. But it's hard for me to find it now. It was a bit confusing because sometimes it wasn't only used for new documents, that's why people asked about it.

**Comment by pryrt** [comment]
@rdipardo said in https://github.com/notepad-plus-plus/notepad-plus-plus/pull/12236#issuecomment-1267700466, 
> Like I said in #12256, a future option to prefer "Default Language" over "Content detected" is going to be more complicated than just a single check box; there are at least two cases with different outcomes:

You keep on gravitating to that.  That's not the option that I described here -- it is not my intention at all,  The option described here is _solely_ whether Default Language or Normal (L_TEXT) applies to existing files that are opened and cannot be determined by File Extension or Content Detected.  I had hoped that @ArkadiuszMichalski's description was understood that way by you six days ago, but it still seems to be how you think of this request.

**Comment by rdipardo** [comment]
> The option described here is solely whether Default Language or Normal (L_TEXT) applies to existing files that are opened and cannot be determined by File Extension or Content Detected.

I still don't see why the automatic highlighting of `Code::Blocks` files should roam free while bash scripts that look like XML is worth preserving. Imagine a bulk of XML just under the 200MB limit with an oddball extension; it's happened already: [XML declaration in file causes Notepad++ to hang for 10s on opening it](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/11319).

> I had hoped that @ArkadiuszMichalski's description was understood that way by you six days ago

What I see is a well-intentioned request for a feature enhancement. The only problem is it should have been put forward when #11504 appeared, four months before I undertook a solution that evidently wasn't wanted. After studying the code, it would be an overstatement to call the lexer selection business a "feature". It's a tangle of `if` blocks thrown against the wall several years ago, and tolerated by most users because its accidental effects are usually beneficial. 

The goal of this new feature is to make a known bug deliberately reproducible. Why not simply reject #11504 as not a bug and say "That's how it works"? There was plenty of time. The committed patch is now an obstacle to be worked around. At least the preliminary labour of identifying the operative decision points is done, so a new option variable can guard the right places. Fortunately it won't be my problem.

**Comment by pryrt** [comment]
I am sorry.  

The reason I didn't put this second request in in 4 months ago was because I didn't think of it as at all four months ago; it was only the recent discussion in the #11504 fix that made me think of this as a compromise (I thought it would help with the problems being discussed, but it obviously didn't), and I thought it might make other users happy who have requested such things in the past.  If I had known it would cause more difficulty and frustration, I wouldn't have made this secondary suggestion.

If anyone wants to close this request, I don't care.

**Comment by donho** [comment]
Let's see if there are other users need it.
